src/libraries/System.Private.CoreLib/src/System/Type.cs

```diff
@@ -134,21 +144,38 @@ public virtual Type[] GetGenericParameterConstraints()
             }
             return GetConstructorImpl(bindingAttr, binder, callConvention, types, modifiers);
         }
+
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)]
         protected abstract ConstructorInfo? GetConstructorImpl(BindingFlags bindingAttr, Binder? binder, CallingConventions callConvention, Type[] types, ParameterModifier[]? modifiers);
 
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)]
         public ConstructorInfo[] GetConstructors() => GetConstructors(BindingFlags.Public | BindingFlags.Instance);
+
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)]
         public abstract ConstructorInfo[] GetConstructors(BindingFlags bindingAttr);
 
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicEvents)]
         public EventInfo? GetEvent(string name) => GetEvent(name, Type.DefaultLookup);
+
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicEvents | DynamicallyAccessedMemberTypes.NonPublicEvents)]
         public abstract EventInfo? GetEvent(string name, BindingFlags bindingAttr);
 
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicEvents)]
         public virtual EventInfo[] GetEvents() => GetEvents(Type.DefaultLookup);
+
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicEvents | DynamicallyAccessedMemberTypes.NonPublicEvents)]
         public abstract EventInfo[] GetEvents(BindingFlags bindingAttr);
 
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicFields)]
```

Comment:
Does the linker remove fields that aren't accessed?  How does that work for things like P/Invokes, where the layout needs to match the native side?  Are such types special-cased?

Reply:
Unless the type has explicit/sequential layout, removing fields is fair game. Stuff used in p/invokes needs to be explicit/sequential.

Reply:
> Stuff used in p/invokes needs to be explicit/sequential.

The default for structs is that they're implicitly sequential, and as a result many structs used in P/Invokes don't explicitly add sequential.  Seems problematic that the linker would remove them.

Reply:
There's no difference between

```csharp
[StructLayout(LayoutKind.Sequential]
struct Foo { }
```

and 

```csharp
struct Foo { }
```

At the IL level.

Structs only have (instance) fields trimmed only if one explicitly puts Auto layout on them.

Reply:
> There's no difference between

Right

> Structs only have (instance) fields trimmed only if one explicitly puts Auto layout on them.

Ah, great.  I think I read your comment "Unless the type has explicit/sequential layout" too fast... I thought you were saying you had to explicitly use sequential layout.